src/linked_hash_map.rs

```diff
@@ -1677,6 +1749,173 @@ impl<'a, K, V> Drop for Drain<'a, K, V> {
     }
 }
 
+impl<'a, K, V, S> CursorMut<'a, K, V, S> {
+    /// Returns an `Option` of the current element in the list, provided it is not the
+    /// _guard_ node, and `None` overwise.
+    #[inline]
+    pub fn current(&mut self) -> Option<(&K, &mut V)> {
+        unsafe {
+            let at = NonNull::new_unchecked(self.cur);
+            self.peek(|| at)
+        }
+    }
+
+    /// Retrieves the next element in the list (moving towards the end).
+    #[inline]
+    pub fn peek_next(&mut self) -> Option<(&K, &mut V)> {
+        unsafe {
+            let at = (*self.cur).links.value.next;
+            self.peek(|| at)
+        }
+    }
+
+    /// Retrieves the previous element in the list (moving towards the front).
+    #[inline]
+    pub fn peek_prev(&mut self) -> Option<(&K, &mut V)> {
+        unsafe {
+            let at = (*self.cur).links.value.prev;
+            self.peek(|| at)
+        }
+    }
+
+    // Retrieves the element specified by the at closure function without advancing current
+    // position to it.
+    #[inline]
+    fn peek(&mut self, at: impl FnOnce() -> NonNull<Node<K, V>>) -> Option<(&K, &mut V)> {
+        if let Some(values) = self.values {
+            unsafe {
+                let node = at().as_ptr();
+                if node == values.as_ptr() {
+                    None
+                } else {
+                    let entry = (*node).entry_mut();
+                    Some((&entry.0, &mut entry.1))
+                }
+            }
+        } else {
+            None
+        }
+    }
+
+    /// Updates the pointer to the current element to the next element in the
+    /// list (that is, moving towards the end).
+    #[inline]
+    pub fn move_next(&mut self) {
+        let at = unsafe { (*self.cur).links.value.next };
+        self.muv(|| at);
+    }
+
+    /// Updates the pointer to the current element to the previous element in the
+    /// list (that is, moving towards the front).
+    #[inline]
+    pub fn move_prev(&mut self) {
+        let at = unsafe { (*self.cur).links.value.prev };
+        self.muv(|| at);
+    }
+
+    /// Positions the cursor at the element associated with the specified key. Returns `true`
+    /// if the element exists and the operation succeeds, or `false` if the element does not
+    /// exist.
+    #[inline]
+    pub fn move_at(&mut self, key: &K) -> bool
+    where
+        K: Eq + Hash,
+        S: BuildHasher,
+    {
+        unsafe {
+            let hash = hash_key(self.hash_builder, key);
+            let i_entry = self
+                .table
+                .find_entry(hash, |o| (*o).as_ref().key_ref().eq(key));
+
+            match i_entry {
+                Ok(occupied) => {
+                    let at = *occupied.get();
+                    self.muv(|| at);
+                    true
+                }
+                Err(_) => false,
+            }
+        }
+    }
```

Comment:
We can skip it for now. If we have the original map in scope, we can always create it later. However, without the original map in scope, it might pose a problem. WDYT?

> It would be amazing if we could convert the cursor back into a `&mut LinkedHashMap` too, but that's hard to do with the way the code is currently and that's not necessary for the first version.

Absolutely, I've been considering the best approach here. However, I paused before the PR review to avoid making too many assumptions on my own.

Reply:
> We can skip it for now. If we have the original map in scope, we can always create it later. However, without the original map in scope, it might pose a problem. WDYT?

If it's a big problem I'd rather be able to go from the cursor back to a `&mut LinkedHashMap`. This suggestion isn't out of nowhere, it was also made for the current `LinkedList` `CursorMut` API: https://github.com/rust-lang/rust/issues/58533#issuecomment-1367564834

Reply:
> I'd rather be able to go from the cursor back to a &mut LinkedHashMap

@kyren, I'm keen to implement this, as it's required for my use case. Could you advise on the best approach for the implementation?

Reply:
From my understanding, it seems we can't use `&mut LinkedHashMap -> CursorMut -> &mut LinkedHashMap` due to the double mutable/exclusive borrowing. The only feasible solution appears to be a consuming conversion: `LinkedHashMap -> CursorMut -> LinkedHashMap`.

Could the `move_at` method, base on that API, potentially become expensive due to frequent re-allocation caused by this consuming API? Is my understanding correct?

Reply:
No, a consuming LinkedHashMap -> CursorMut -> LinkedHashMap API is not necessary at all, the API of `CursorMut` -> `&mut LinkedHashMap` is *reborrowing*, the only requirement is to refactor `CursorMut` to contain a `&mut LinkedHashMap` instead of individually borrowing all four fields from the parent `LinkedHashMap`.

Reply:
It would be pretty straightforward using the raw table API rather than the normal hash_table API. I don't know *how bad* it would be to depend on the `raw` feature of `hashbrown`, but considering that the raw API is the normal implementation of hashbrown made public, I doubt it's *that bad*.

Reply:
After further consideration, I'm not fully convinced that we're at the right direction with this. Let me elaborate.

The primary objective of the `move_at` operation is to enable efficient cursor relocation using a key while keep this operation cost-effective as it may be part of intensive computations, in my case it is.

Considering the `&mut hashlink::OccupiedEntry` -> `CursorMut` case, when we use `CursorMut` -> `&mut LinkedHashMap` conversion for cursor's repositioning, it can be an expensive compute even on stack allocation.

I'm not entirely sure how the Rust compiler handles this case. Could you share your insights?

Reply:
> Considering the &mut hashlink::OccupiedEntry -> CursorMut case, when we use CursorMut -> &mut LinkedHashMap conversion for cursor's repositioning, it can be an expensive compute even on stack allocation.

This would be *free*?

Reply:
Apologies if my previous message wasn't clear. I'm trying to understand if a `CursorMut` is created on the stack each time I convert `&mut hashlink::OccupiedEntry` to `CursorMut` (without using `&mut CursorMut`). Is that correct?

Reply:
> 'm trying to understand if a CursorMut is created on the stack each time I convert &mut hashlink::OccupiedEntry to CursorMut

You're asking me if when you create a new `CursorMut`, that's created on the stack? Yes? How else could it work?

Reply:
As far as I can see, there's no cost argument at all to be made here, if that's what you're trying to argue. A `CursorMut::move_at` method is a weird method to provide *because* it's not any cheaper to do that than to find a new entry and make a cursor out of it, since both of them do the same thing (find an entry from scratch and get the linked list pointer). You're presenting an API to the user that makes it seem as though there's some kind of benefit when there's actually not.

It is currently easier to provide `CursorMut::move_at` due to basically an accident of the hashbrown API and mutable borrowing works, but you can make both `CursorMut` *and* `RawOccupiedEntryMut` smaller *and* provide a strictly more flexible API by using the hashbrown raw API with bucket pointers, sidestepping the mutable borrow problem. This is also incidentally exactly the same way hashbrown itself works, if you look at the `hash_table::OccupiedEntry`, you'll see that it contains a borrow of the parent table and a bucket pointer, which is exactly what I'm proposing: https://docs.rs/hashbrown/latest/src/hashbrown/table.rs.html#1447-1454.

Reply:
> You're asking me if when you create a new CursorMut, is it created on the stack? Yes? How else could it work?

No, not really. Could you please clarify if there's a performance difference between the two hypothetical cases:

```rust
// 1
let mut cursor = map.cursor_front_mut();
for i in 0..1_000_000_000 {
    match cursor.move_at(&i) {
        Some(_) => cursor.insert_before(new_k, new_v),
        None => todo!(),
    };
};
```

```rust
// 2
for i in 0..1_000_000_000 {
    match map.raw_entry_mut().from_key(&i) {
        RawEntryMut::Occupied(occupied) => {
            let mut cursor = occupied.cursor_mut(); // a new instance of CursorMut has been created
            cursor.insert_before(new_k, new_v);
        },
        RawEntryMut::Vacant(_) => todo!(),
    };
};
```

In the second scenario, we generate a cursor each time we access the `RawEntryMut::Occupied(occupied) =>` code path. I'm curious if this could impact performance compared to the first scenario where we reuse an existing cursor. Apologies if this question seems basic - I'm still getting acquainted with how the Rust compiler optimizes this.

Kind regards,
Oleg

Reply:
Speaking practically, there is **no** performance difference between the two. `move_at` and `from_key` both have to do the same *work*, which is to hash the given key and call `HashTable::find` to get the pointer to the linked list. Thinking about "creating a `CursorMut`" itself as having a cost, or trying to "reuse an existing `CursorMut`" is not usually something that Rust programmers even think about, it's a *pointer* and a couple of references, optimizing compilers eat this stuff for breakfast. This is why I was so confused before.

Besides this, in my proposed change, the number of references would go from four to one, and then the *only* thing that would happen is that you would be throwing away the other field `cur: *mut Node<K, V>` and then recreating it by doing another call to `HashTable::find`, **which is what happens either way**. But even this is most likely not going to have much of a (good) performance impact because again, optimizing compilers eat this stuff for breakfast, though it might give the optimizer slightly less work to do.

In some situations there is a *very small* potential benefit has nothing to do with "reusing" a simple stack variable, it's that creating a `CursorMut` has to call `ensure_guard_node`. This doesn't apply to your example though, because you're going through `RawOccupiedEntryMut` which does not have to do this call. Even if you do end up creating a `CursorMut` from `HashMap::cursor_mut()` (which does have to call `ensure_guard_node`) in a loop, the `ensure_guard_node` call is almost certainly going to be lifted out of the loop because the compiler will be able to see that `self.values` gets checked to make sure it's allocated but is never set to NULL within the loop.

Speaking more precisely, optimizing compilers are hard to predict and it's always possible that one or the other loop will be compiled differently, but I don't know which one would be faster or how they'd be different, and if they are measurably different stuff like this is usually considered an "optimization bug". My instincts tell me actually that the worst part of `CursorMut` is having four references instead of one and that this is worse than any `ensure_guard_node` call, but it's honestly just a wild guess. When you get down to these sorts of micro-optimizations, usually you just do the simplest, clearest thing unless you have a lot of *good* evidence to do something else. The API of a library is the primary way that the developer of that library communicates with its users, and having an API which appears to re-use a `CursorMut` even though there's *nothing to re-use* feels like a **lie**, and I feel like the default decision is not to provide that API unless there's some *really clear* benefit to it.

I'm also emphatically *not* asking you to do the work of converting to using the hashbrown raw API either, in fact I'm not really in a terrible rush to do this at all (not because of any *real* downside, but because the raw API is described scarily in terms of *stability* and as this crate is somewhat foundational, I'm cautious about writing ourselves into a corner). In fact, the only way that either `move_at` or `into_map` are useful at *all* is in the rare situation where you have a `CursorMut` but *not* the original map (such as in a utility function that takes a `CursorMut` as a parameter), you can always just *drop the `CursorMut` and re-create it* after all. This discussion makes me think that this is actually the situation you're in, but you've convinced yourself that re-using a `CursorMut` is helpful in some way when it's almost certainly not.

Does this clear things up?

Reply:
> I'm also emphatically not asking you to do the work of converting to using the hashbrown raw API either, in fact I'm not really in a terrible rush to do this at all (not because of any real downside, but because the raw API is described scarily in terms of stability and as this crate is somewhat foundational, I'm cautious about writing ourselves into a corner if it's removed). In fact, the only way that either move_at or into_map are useful at all is in the rare situation where you have a CursorMut but not the original map (such as in a utility function that takes a CursorMut as a parameter), you can always just drop the CursorMut and re-create it after all. 

I don't mind to give the proposed impl a try. I'm unsure yet about how much work it's required though, but using an experimental API doesn't sound ideal to me either. I'd prefer to exclude very rare cases from the library's scope, especially if it's functionality is reachable via existing API with no performance penalty.

> This discussion makes me think that this is actually the situation you're in, but you've convinced yourself that re-using a CursorMut is helpful in some way when it's almost certainly not.

You're right. I was about to prevent a scenario where I need to write _some_ boilerplate code for the move-at to functionality, even though it could technically work without it, as you say. 

I initially considered it _a helper_ (and this is what I mentioned in my removed comment), but later I've got concerns about the performance impact of not having it. However, your explanation has alleviated those concerns. I trust your judgement on this. 

> Does this clear things up?

Yep, everything is clear now; thanks for your patience.